src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -0,0 +1,629 @@
+        self._reference = reference
+        # Store the absolute position for later.
+        self._reference_abs_pos = reference.abs_position
+        # To enable loops to work correctly we can set the start/stop point
+        # and not just use base it on the reference's absolute position
+        if start_point and not isinstance(start_point, int):
+            raise TypeError(f"The start_point passed into a "
+                            f"DefinitionUseChain must be an int but found "
+                            f"'{type(start_point).__name__}'.")
+        if stop_point and not isinstance(stop_point, int):
+            raise TypeError(f"The stop_point passed into a "
+                            f"DefinitionUseChain must be an int but found "
+                            f"'{type(stop_point).__name__}'.")
+        self._start_point = start_point
+        self._stop_point = stop_point
+        if control_flow_region is None:
+            self._scope = [reference.ancestor(Routine)]
+            if self._scope[0] is None:
+                self._scope = reference.root.children[:]
+        else:
+            # We need a list of regions for control flow.
+            if not isinstance(control_flow_region, list):
+                raise TypeError(
+                    f"The control_flow_region passed into a "
+                    f"DefinitionUseChain must be a list but found "
+                    f"'{type(control_flow_region).__name__}'."
+                )
+            self._scope = control_flow_region
+        if is_local is None:
+            # TODO Work out if symbol is local to the scope or
+            # globally accessible. Anything that isn't a local
+            # variable in a Routine has to be assumed as globally
+            # accessible for now.
+            self._is_local = isinstance(
+                reference.symbol.interface, AutomaticInterface
+            )
+        else:
+            self._is_local = is_local
+
+        # The uses, defsout and killed sets as defined for each basic block.
+        self._uses = []
+        self._defsout = []
+        self._killed = []
+
+        # The output map, mapping between nodes and the reach of that node.
+        self._reaches = []
+        self._writes = []
+
+    @property
+    def uses(self):
+        """
+        :returns: the list of nodes used by this DefinitionUseChain.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._uses
+
+    @property
+    def defsout(self):
+        """
+        :returns: the list of output nodes computed by this DefinitionUseChain.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._defsout
+
+    @property
+    def killed(self):
+        """
+        :returns: the list of killed output nodes computed by this
+                  DefinitionUseChain.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._killed
```

Comment:
Can you expand what defsout are, I don't quite understand what they are and how are they different than killed?

Reply:
Ok - so the implementation does need/use the killed accesses now, I need some more tests for behviour inside control flow for codeblocks/calls to ensure this is behaving as I expect.

Reply:
So for defout can we use something like:
`:returns: the list of nodes that reach the end of the block without being killed, and therefore can have dependencies outside of this block."

And for killed:
`:returns: the list of nodes that represent the last use of an assigned variable. The next access is a write that reassigns it's value."

Again, supposing that these are the right descriptions?

Reply:
I changed to be similar to this, I think these are ok now (though I changed the killed slightly from what you put).